File: graph.cs
// 重み付き(重みなし)の無向グラフを管理する.
// Depends on: Edge<T>, UnionFind
// @author Nauclhlt.
public sealed class Graph<T> where T : struct, INumber<T>
{
    private List<List<Edge<T>>> _graph;
    private List<Edge<T>> _edges;
    private bool[] _seen;
    private UnionFind _uf;

    private int _vertexCount;

    public int VertexCount => _vertexCount;
    public bool[] Seen => _seen;
    public UnionFind UnionFind => _uf;
    public List<List<Edge<T>>> RawGraph => _graph;
    public List<Edge<T>> Edges => _edges;

    public Graph(int vertexCount)
    {
        _vertexCount = vertexCount;
        _graph = new(_vertexCount);
        _edges = new List<Edge<T>>(_vertexCount);
        for (int i = 0; i < _vertexCount; i++)
        {
            _graph.Add(new());
        }
    }

    // 頂点fromとtoの間に重みweightの無向辺を追加する.
    // O(1)
    public void AddEdge(int a, int b, T weight)
    {
        if (!Validate(a) || !Validate(b)) return;

        if (a > b) (a, b) = (b, a);

        _graph[a].Add(new Edge<T>(a, b, weight));
        _graph[b].Add(new Edge<T>(b, a, weight));
        _edges.Add(new Edge<T>(a, b, weight));
    }

    // このグラフの連結の状態を表すDSUを生成する.
    // O(V+E)
    public void SetupDSU()
    {
        _uf = new UnionFind(_vertexCount);
        for (int i = 0; i < _edges.Count; i++)
        {
            _uf.Unite(_edges[i].From, _edges[i].To);
        }
    }

    // 探索用の配列を作成する.
    public void SetupSearch()
    {
        _seen = new bool[_vertexCount];
    }

    // 頂点a, bが同じ連結成分に属するかを判定する.
    // ほぼ定数時間
    public bool Same(int a, int b)
    {
        if (_uf is null) throw new Exception("Call SetupSearch.");

        return _uf.Same(a, b);
    }

    // 連結成分ごとに頂点のリストを返す.
    // O(V)
    public Dictionary<int, List<int>> GetConnectedComponents()
    {
        if (_uf is null) throw new Exception("Call SetupSearch.");
        return _uf.FindAll();
    }

    // 頂点nからDijkstra法を用いてそれぞれの頂点に対する最短経路を求める.
    // O(VlogE)
    public void DijkstraFrom(int n, T[] map)
    {
        if (!Validate(n)) return;

        if (_seen is null) throw new Exception("call SetupSearch.");

        Array.Clear(_seen);

        map[n] = T.Zero;

        PriorityQueue<int, T> pq = new();

        pq.Enqueue(n, T.Zero);

        while (pq.Count > 0)
        {
            int p = pq.Dequeue();

            if (_seen[p]) continue;

            _seen[p] = true;

            List<Edge<T>> children = _graph[p];
            for (int i = 0; i < children.Count; i++)
            {
                T w = map[p] + children[i].Weight;
                if (w < map[children[i].To])
                {
                    map[children[i].To] = w;
                    pq.Enqueue(children[i].To, map[children[i].To]);
                }
            }
        }
    }

    // 頂点nからBFSをしてそれぞれの頂点にはじめに訪問したときのパスの重みの合計を配列に格納する.
    // O(V+E)
    public void BfsFrom(int n, T[] map)
    {
        if (!Validate(n)) return;

        if (_seen is null) throw new Exception("call SetupSearch.");

        Array.Clear(_seen);

        map[n] = T.Zero;

        Queue<(int, T)> queue = new();

        queue.Enqueue((n, T.Zero));

        while (queue.Count > 0)
        {
            (int p, T w) = queue.Dequeue();

            if (_seen[p]) continue;

            _seen[p] = true;
            map[p] = w;

            List<Edge<T>> children = _graph[p];
            for (int i = 0; i < children.Count; i++)
            {
                queue.Enqueue((children[i].To, w + children[i].Weight));
            }
        }
    }

    // 補グラフを作成する.
    // O(V^2)
    public Graph<T> CreateComplement()
    {
        HashSet<(int, int)> edgeSet = new();
        for (int i = 0; i < _edges.Count; i++)
        {
            edgeSet.Add((_edges[i].From, _edges[i].To));
        }

        Graph<T> g = new(_vertexCount);

        for (int i = 0; i < _vertexCount - 1; i++)
        {
            for (int j = i + 1; j < _vertexCount; j++)
            {
                if (!edgeSet.Contains((i, j)))
                {
                    g.AddEdge(i, j, default);
                }
            }
        }

        return g;
    }

    // 二部グラフか判定する.
    // O(V+E)
    public bool IsBipartite()
    {
        if (_seen is null) throw new Exception("Call SetupSearch.");

        Array.Clear(_seen);

        Stack<(int, bool)> stack = new();

        bool[] memo = new bool[_vertexCount];

        for (int i = 0; i < _vertexCount; i++)
        {
            stack.Push((i, false));

            while (stack.Count > 0)
            {
                (int n, bool c) = stack.Pop();

                if (_seen[n])
                {
                    if (memo[n] != !c) return false;
                    continue;
                }

                _seen[n] = true;
                memo[n] = !c;

                var ch = _graph[n];
                for (int j = 0; j < ch.Count; j++)
                {
                    stack.Push((ch[j].To, !c));
                }
            }
        }

        return true;
    }

    // 木の直径を求める.
    // O(V+E)
    public T TreeDiameter()
    {
        if (_seen is null) throw new Exception("call SetupSearch.");
        if (_edges.Count != _vertexCount - 1)
        {
            throw new Exception("Not a tree graph.");
        }

        T[] dist = new T[_vertexCount];

        BfsFrom(0, dist);

        T max = T.Zero;
        int v = 0;
        for (int i = 0; i < _vertexCount; i++)
        {
            if (dist[i] > max)
            {
                max = dist[i];
                v = i;
            }
        }

        BfsFrom(v, dist);

        return dist.Max();
    }

    // Kruskal法で最大全域木の重みの総和を求める.
    // O(V+E)
    public T MaxSpanningTreeWeight()
    {
        UnionFind unionFind = new(_vertexCount);

        T ans = T.Zero;
        foreach (var edge in _edges.OrderByDescending(x => x.Weight))
        {
            if (!unionFind.Same(edge.From, edge.To))
            {
                unionFind.Unite(edge.From, edge.To);
                ans += edge.Weight;
            }
        }

        return ans;
    }

    // Kruskal法で最小全域木の重みの総和を求める
    // O(V+E)
    public T MinSpanningTreeWeight()
    {
        UnionFind unionFind = new(_vertexCount);

        T ans = T.Zero;
        foreach (var edge in _edges.OrderBy(x => x.Weight))
        {
            if (!unionFind.Same(edge.From, edge.To))
            {
                unionFind.Unite(edge.From, edge.To);
                ans += edge.Weight;
            }
        }

        return ans;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    bool Validate(int n)
    {
        return 0 <= n && n < _vertexCount;
    }
}

File: cpp/graph.cpp
template <typename T>
class Graph
{
private:
    vector<vector<Edge<T>>> _graph;
    vector<Edge<T>> _edges;
    vector<bool> _seen;
    unique_ptr<UnionFind> _uf;
    int _vertexCount;

public:
    Graph(int vertexCount)
    {
        _vertexCount = vertexCount;
        _graph.resize(_vertexCount);
        _edges.reserve(_vertexCount);
        _seen.resize(_vertexCount);
        _uf = NULL;
    }

    int VertexCount()
    {
        return _vertexCount;
    }

    vector<vector<Edge<T>>>& RawGraph()
    {
        return _graph;
    }

    vector<Edge<T>>& Edges()
    {
        return _edges;
    }

    UnionFind DSU()
    {
        return *_uf;
    }

    void AddEdge(int a, int b, T weight)
    {
        if (!Validate(a) || !Validate(b))
        {
            return;
        }

        if (a > b)
        {
            swap(a, b);
        }

        _graph[a].push_back(Edge<T>(a, b, weight));
        _graph[b].push_back(Edge<T>(b, a, weight));
        _edges.push_back(Edge<T>(a, b, weight));
    }

    void SetupDSU()
    {
        _uf.reset(new UnionFind(_vertexCount));
        for (int i = 0; i < (int)_edges.size(); i++)
        {
            _uf->Unite(_edges[i].From, _edges[i].To);
        }
    }

    bool Same(int a, int b)
    {
        if (_uf == nullptr)
        {
            throw exception();
        }
        return _uf->Same(a, b);
    }

    unordered_map<int, vector<int>> GetConnectedComponents()
    {
        if (_uf == nullptr)
        {
            throw exception();
        }
        return _uf->FindAll();
    }

    void DijkstraFrom(int n, vector<T>& map)
    {
        if (!Validate(n))
        {
            return;
        }

        fill(_seen.begin(), _seen.end(), false);

        fill(map.begin(), map.end(), numeric_limits<T>::max());
        map[n] = 0;

        priority_queue<pair<T, int>, vector<pair<T, int>>, greater<pair<T, int>>> pq;
        pq.emplace(0, n);

        while (!pq.empty())
        {
            T c = pq.top().first;
            int p = pq.top().second;
            pq.pop();

            if (_seen[p])
            {
                continue;
            }

            _seen[p] = true;

            vector<Edge<T>>& ch = _graph[p];
            for (int i = 0; i < (int)ch.size(); i++)
            {
                T w = map[p] + ch[i].Weight;
                if (w < map[ch[i].To])
                {
                    map[ch[i].To] = w;
                    pq.emplace(w, ch[i].To);
                }
            }
        }
    }

    vector<vector<T>> WarshallFloyd()
    {
        if (_vertexCount > 800)
        {
            throw exception();
        }

        T inf = numeric_limits<T>::max();

        vector<vector<T>> map(_vertexCount, vector<T>(_vertexCount, inf));

        for (int i = 0; i < _vertexCount; i++)
        {
            map[i][i] = 0;
        }

        for (int i = 0; i < (int)_edges.size(); i++)
        {
            map[_edges[i].From][_edges[i].To] = min(_edges[i].Weight, map[_edges[i].From][_edges[i].To]);
            map[_edges[i].To][_edges[i].From] = min(_edges[i].Weight, map[_edges[i].To][_edges[i].From]);
        }

        for (int k = 0; k < _vertexCount; k++)
        {
            for (int j = 0; j < _vertexCount; j++)
            {
                for (int i = 0; i < _vertexCount; i++)
                {
                    if (map[i][k] != inf && map[k][j] != inf)
                    {
                        if (map[i][k] + map[k][j] < map[i][j])
                        {
                            map[i][j] = map[i][k] + map[k][j];
                        }
                    }
                }
            }
        }

        return map;
    }

    void BfsFrom(int n, vector<T>& map)
    {
        if (!Validate(n))
        {
            return;
        }

        fill(_seen.begin(), _seen.end(), false);

        map[n] = 0;

        queue<pair<int, T>> queue;

        queue.emplace(n, 0);

        while (!queue.empty())
        {
            int p = queue.front().first;
            T w = queue.front().second;
            queue.pop();

            if (_seen[p])
                continue;

            _seen[p] = true;
            map[p] = w;

            vector<Edge<T>>& ch = _graph[p];
            for (int i = 0; i < (int)ch.size(); i++)
            {
                queue.emplace(ch[i].To, w + ch[i].Weight);
            }
        }
    }

    Graph<T> CreateComplement() 
    {
        struct pair_hash {
            inline std::size_t operator()(const std::pair<int,int> & v) const {
                return v.first ^ v.second;
            }
        };

        unordered_set<pair<int, int>, pair_hash> edgeSet;
        for (int i = 0; i < (int)_edges.size(); i++)
        {
            edgeSet.emplace(_edges[i].From, _edges[i].To);
        }

        Graph<T> g(_vertexCount);

        for (int i = 0; i < _vertexCount - 1; i++)
        {
            for (int j = i + 1; j < _vertexCount; j++)
            {
                if (edgeSet.find(make_pair(i, j)) == edgeSet.end())
                {
                    cout << i << ", " << j << endl;
                    g.AddEdge(i, j, 0);
                }
            }
        }

        return g;
    }
    
    bool IsBipartite()
    {
        fill(_seen.begin(), _seen.end(), false);

        stack<pair<int, bool>> stack;

        vector<bool> memo(_vertexCount);

        for (int i = 0; i < _vertexCount; i++)
        {
            if (_seen[i]) continue;

            stack.emplace(i, false);

            while (!stack.empty())
            {
                int n = stack.top().first;
                bool c = stack.top().second;
                stack.pop();

                if (_seen[n])
                {
                    if (memo[n] != c)
                    {
                        return false;
                    }

                    continue;
                }

                _seen[n] = true;
                memo[n] = c;

                vector<Edge<T>>& ch = _graph[n];
                for (int j = 0; j < (int)ch.size(); j++)
                {
                    stack.emplace(ch[j].To, !c);
                }
            }
        }

        return true;
    }

    T TreeDiameter()
    {
        if ((int)_edges.size() != _vertexCount - 1)
        {
            throw exception();
        }

        vector<T> dist(_vertexCount);

        BfsFrom(0, dist);

        T max = 0;
        int v = 0;
        for (int i = 0; i < _vertexCount; i++)
        {
            if (dist[i] > max)
            {
                max = dist[i];
                v = i;
            }
        }

        fill(dist.begin(), dist.end(), 0);

        BfsFrom(v, dist);

        max = 0;
        for (int i = 0; i < _vertexCount; i++)
        {
            if (dist[i] > max)
            {
                max = dist[i];
            }
        }

        return max;
    }

    T MaxSpanningTreeWeight()
    {
        UnionFind unionFind(_vertexCount);

        T ans = 0;

        auto cmp = [](const Edge<T> &a, const Edge<T> &b)
        {
            return a.Weight > b.Weight;
        };

        vector<Edge<T>>& edges = _edges;

        sort(edges.begin(), edges.end(), cmp);

        for (int i = 0; i < (int)edges.size(); i++)
        {
            if (!unionFind.Same(edges[i].From, edges[i].To))
            {
                unionFind.Unite(edges[i].From, edges[i].To);
                ans += edges[i].Weight;
            }
        }

        return ans;
    }

    T MinSpanningTreeWeight()
    {
        UnionFind unionFind(_vertexCount);

        T ans = 0;

        auto cmp = [](const Edge<T> &a, const Edge<T> &b)
        {
            return a.Weight < b.Weight;
        };

        vector<Edge<T>>& edges = _edges;

        sort(edges.begin(), edges.end(), cmp);

        for (int i = 0; i < (int)edges.size(); i++)
        {
            if (!unionFind.Same(edges[i].From, edges[i].To))
            {
                unionFind.Unite(edges[i].From, edges[i].To);
                ans += edges[i].Weight;
            }
        }

        return ans;
    }

private:
    inline bool Validate(int n)
    {
        return 0 <= n && n < _vertexCount;
    }
};